lib/models/per-bundle-addon-cache/index.js

```diff
@@ -0,0 +1,220 @@
+    let bundleRootCache = this.bundleRootCaches.get(bundleRootPkgInfo);
+
+    let cacheEntry = bundleRootCache.get(addonPkgInfo);
+    if (cacheEntry) {
+      cacheEntry.instance = addonInstance;
+    } else {
+      cacheEntry = new CacheEntry(addonInstance);
+      bundleRootCache.set(addonPkgInfo, cacheEntry);
+    }
+  }
+
+  /**
+   * Retrieve the addon instance for the given addonPkgInfo within the cache for the
+   * given bundleRootPkgInfo
+   *
+   * @public
+   * @method getInstance
+   * @param {String} bundleRootPkgInfo the packageInfo for the bundle root
+   * @param {String} addonPkgInfo the packageInfo for the addon instance desired
+   * @return {Object} the addon instance for the given addonPkgInfo in the bundle root's cache,
+   * or null/undefined if no entry exists or the instance is not set.
+   */
+  getPerBundleInstance(bundleRootPkgInfo, addonPkgInfo) {
+    let bundleRootCache = this.bundleRootCaches.get(bundleRootPkgInfo);
+    if (!bundleRootCache) {
+      // Normally this should not be possible, if the bundleRootPkgInfo really is for
+      // a bundle root. But just in case...
+      return null;
+    }
+
+    let cacheEntry = bundleRootCache.get(addonPkgInfo);
+    return cacheEntry ? cacheEntry.instance : null;
+  }
+
+  /**
+   * Retrieve the cache entry for the given addonPkgInfo in the cache for a particular bundleRootPkgInfo.
+   *
+   * @public
+   * @method getPerBundleCacheEntry
+   * @param {String} bundleRootPkgInfo the packageInfo for the bundle root
+   * @param {String} addonPkgInfo the packageInfo for the addon instance desired
+   * @return {CacheEntry} the cache entry for the addon, or null
+   */
+  getPerBundleCacheEntry(bundleRootPkgInfo, addonPkgInfo) {
+    let bundleRootCache = this.bundleRootCaches.get(bundleRootPkgInfo);
+    if (!bundleRootCache) {
+      return null;
+    }
+
+    let cacheEntry = bundleRootCache.get(addonPkgInfo);
+    return cacheEntry;
+  }
+
+  /**
+   * Indicate if an entry for the given addonPkgInfo exists in the cache for a given
+   * bundleRootPkgInfo (whether or not the actual instance has been set)
+   *
+   * @public
+   * @method hasPerBundleCacheEntry
+   * @param {String} bundleRootPkgInfo the packageInfo for the bundle root
+   * @param {String} addonPkgInfo the packageInfo for the addon instance
+   * @return true if an entry exists, false otherwise.
+   */
+  hasPerBundleCacheEntry(bundleRootPkgInfo, addonPkgInfo) {
+    let bundleRootCache = this.bundleRootCaches.get(bundleRootPkgInfo);
+    if (!bundleRootCache) {
+      return false;
+    }
+
+    let cacheEntry = bundleRootCache.get(addonPkgInfo);
+    return cacheEntry !== null && cacheEntry !== undefined;
+  }
+
+  /**
+   * For use with per-bundle addon caching, is a given object an instance of a Project?
+   * 
+   * @public
+   * @method isProject
+   * @param {Object} obj 
+   * @returns {Boolean}
+   */
+  static isProject(obj) {
+    return obj?.constructor?.name === 'Project' || false;
```

Comment:
Can't use optional chaining, also it seems pretty odd to check the constructor like this. Why wouldn't we use the methods on the project instance?

Reply:
Agree. I was originally just following the 'isLazyEngine' style, but yeah, (obj instanceof Project is obvious and much more straightforward.

Reply:
Removed 'isProject' entirely here as it wasn't needed.